src/main/java/org/opentripplanner/middleware/bugsnag/jobs/BugsnagEventJob.java

```diff
@@ -0,0 +1,113 @@
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.mongodb.client.model.Filters.eq;
+import static org.opentripplanner.middleware.spark.Main.getConfigPropertyAsInt;
+
+/**
+ * This job is responsible for maintaining Bugsnag event data. This is achieved by managing the event request jobs
+ * triggered by {@link BugsnagEventJob}, obtaining event data from Bugsnag storage and removing stale events.
+ *
+ * Event requests triggered by {@link BugsnagEventJob} are not completed immediately, instead a job is initiated by
+ * Bugsnag and a 'pending' event request is returned. This event request is then checked with Bugsnag every minute
+ * until the status becomes 'completed'. At this point the event data compiled by the original request made by
+ * {@link BugsnagEventJob} is available for download from a unique URL now present in the updated event request. This is
+ * downloaded and saved to Mongo. Any event data that is older than the reporting window is then deleted.
+ */
+public class BugsnagEventJob implements Runnable {
+
+    private static final int BUGSNAG_REPORTING_WINDOW_IN_DAYS
+        = getConfigPropertyAsInt("BUGSNAG_REPORTING_WINDOW_IN_DAYS", 14);
+
+    private static TypedPersistence<BugsnagEventRequest> bugsnagEventRequests = Persistence.bugsnagEventRequests;
+    private static TypedPersistence<BugsnagEvent> bugsnagEvents = Persistence.bugsnagEvents;
+
+    /**
+     * On each cycle get the newest event data request from Mongo. If that request has been fulfilled by Bugsnag, add
+     * all new events to Mongo and remove all that are older than the Bugsnag reporting window.
+     */
+    public void run() {
+        Bson filter = Filters.ne("status", "complete");
+        BugsnagEventRequest originalRequest =
+            bugsnagEventRequests.getOneFiltered(filter, Sorts.descending("dateCreated"));
+        if (originalRequest != null) {
+            manageEvents(originalRequest);
+        }
+        removeStaleEvents();
+    }
+
+    /**
+     * Confirm that the event request has completed and update event data accordingly.
+     */
+    private void manageEvents(BugsnagEventRequest originalRequest) {
+
+        BugsnagEventRequest currentRequest = BugsnagDispatcher.getEventDataRequest(originalRequest);
+        if (!currentRequest.status.equalsIgnoreCase("completed")) {
+            // request not completed by Bugsnag yet
+            return;
+        }
+
+        removeStaleEventRequests(originalRequest);
+
+        // get event data produced from original event request from Bugsnag storage
+        List<BugsnagEvent> events = BugsnagDispatcher.getEventData(currentRequest);
+
+        //TODO Potential bottleneck depending on the number of projects and events. Dropping all rows and than
+        // inserting would speed this up, but you run the risk of return nothing to the admin dashboard.
```

Comment:
Databases are funky and sometimes bulk operations may not actually be faster. I think this might be premature optimization anyways, so I'm thinking it's fine to keep this code for now. Let's just combine this comment with the one below it.

Reply:
Updated.